narwhal/primary/src/state_handler.rs

```diff
@@ -185,16 +250,184 @@ impl RandomnessState {
+        self.last_narwhal_round_sent = next_leader_narwhal_round;
+
+        let leader = self.leader_schedule.leader(next_leader_narwhal_round);
+        if self.authority_id == leader.id() {
+            // We're the next leader, no need to send an RPC.
+            self.receive_partial_signatures(self.authority_id, self.randomness_round, sigs)
+                .await;
+            return;
+        }
+
+        let peer_id = anemo::PeerId(leader.network_key().0.to_bytes());
+        let peer = self.network.waiting_peer(peer_id);
+        let mut client = PrimaryToPrimaryClient::new(peer);
+        const SEND_PARTIAL_SIGNATURES_TIMEOUT: Duration = Duration::from_secs(10);
+        let request = anemo::Request::new(SendRandomnessPartialSignaturesRequest {
+            round: self.randomness_round,
+            sigs,
+        })
+        .with_timeout(SEND_PARTIAL_SIGNATURES_TIMEOUT);
+
+        if let Some(task) = &self.partial_sig_sender {
+            // Cancel previous partial signature transmission if it's not yet complete.
+            task.abort();
+        }
+        self.partial_sig_sender = Some(spawn_logged_monitored_task!(
+            async move {
+                let resp = client.send_randomness_partial_signatures(request).await;
+                if let Err(e) = resp {
+                    info!(
+                        "random beacon: error sending partial signatures to leader {leader:?}: {e:?}"
+                    );
+                }
+            },
+            "RandomnessSendPartialSignatures"
+        ));
+    }
+
+    async fn receive_partial_signatures(
+        &mut self,
+        authority_id: AuthorityIdentifier,
+        round: u64,
+        sigs: Vec<RandomnessPartialSignature>,
+    ) {
+        let dkg_output = match &self.dkg_output {
+            Some(dkg_output) => dkg_output,
+            None => {
+                error!("random beacon: called receive_partial_signatures before DKG completed");
+                return;
             }
+        };
+        const MAX_ROUND_DELTA: u64 = 2;
+        if round < self.randomness_round {
+            debug!(
+                "random beacon: ignoring partial signatures for old round {round} (we are at {})",
+                self.randomness_round
+            );
+            return;
+        }
+        if round > self.randomness_round + MAX_ROUND_DELTA {
+            debug!(
+                "random beacon: ignoring partial signatures for too-new round {round} (we are at {})",
+                self.randomness_round
+            );
+            return;
+        }
+        if let Some(last_sent) = self.last_randomness_round_sent {
+            if last_sent >= self.randomness_round {
+                debug!(
+                    "random beacon: ignoring partial signatures for already-finished round {round}"
+                );
+                return;
+            }
+        }
+        if let Err(e) = ThresholdBls12381MinSig::partial_verify_batch(
```

Comment:
Optimistically we can collect enough partial signatures -> try to recover the full signature:
- if ok -> we skipped all the sig verifications (including ones we don't even need for the full sig recovery)
- if not -> we fallback to sig verifications -> remove the invalid ones -> continue collecting, etc
This is what we do for other certificates (e.g., in the quorum driver).

Should we do that here as well? what would be the benefit of checking that as part of the header verification?

Reply:
yes - aggregate currently doesn't check the resulting sig, so call aggregate -> verify the sig -> call partial verify only if it fails

Reply:
Would the idea there be to save on compute, or any other benefits? We can verify partial sigs continuously as they are received over the network, so I think it shouldn't be slowing things down much, plus the current flow seems easier to understand IMO. If we check sigs as part of header verification, then later on if a node is catching up, that node won't need to reverify the randomness sig (because it can trust based on the certificate sigs that 2f+1 nodes already attest the randomness is valid)